### ISSUE TITLE
GLIBCXX_3.4.21 not found > Debian Jessie 8.2

### ISSUE THREAD

**Issue by cmyk** [mapnik/node-mapnik#710]
It installs fine with npm and node 6.9.1.
But when I try to run it, I get:
```
osm@map:~/carto/kosmtik$ node_modules/mapnik/bin/mapnik-index.js 
/home/osm/carto/kosmtik/node_modules/mapnik/lib/binding/node-v48-linux-x64/mapnik-index: /usr/lib/x86_64-linux-gnu/libstdc++.so.6: version `GLIBCXX_3.4.21' not found (required by /home/osm/carto/kosmtik/node_modules/mapnik/lib/binding/node-v48-linux-x64/mapnik-index)
/home/osm/carto/kosmtik/node_modules/mapnik/lib/binding/node-v48-linux-x64/mapnik-index: /usr/lib/x86_64-linux-gnu/libstdc++.so.6: version `GLIBCXX_3.4.21' not found (required by /home/osm/carto/kosmtik/node_modules/mapnik/lib/binding/node-v48-linux-x64/libmapnik.so)
```

**Comment by cmyk** [comment]
What version of node is recommended?
I just tried with 4.3.0 and I could install 3.5.13.

**Comment by brendanvinson** [comment]
Your server's c++ compiler is not a high enough version to support 3.5.14, rather than a specific node version you should target.

I ran into this trying to run on CentOS 7 and ended up switching to Ubuntu Xenial instead to solve it.

**Comment by forsetius** [comment]
Same error on LMDE2. Both LMDE2 and Jessie have the same symbols available - `GLIBCXX_3.4.20` being highest. Problem is I have to deploy on Jessie so I can't just switch to Xenial. Also I need Mapnik 3.5.14 because of support for Node 6 - Node 4 isn't supported anymore.

Also, using http://unix.stackexchange.com/questions/284817/how-to-install-gcc-5-on-debian-jessie-8-1/290021 I tried to add the repo for Debian Testing and install only `libstdc++-5-dev` but turns out I cannot do that too - apt says that it needs to uninstall several packages, among them `postgres9.4-postgis2.1`

What can I do? Is there a way to install needed packages in separate directory and use them only for Mapnik? Does Mapnik need to be recompiled? Please provide some instructions if there is any way to do it. Thanks in advance.

**Comment by pnorman** [comment]
Probably caused by the #660 bug

**Comment by aaliddell** [comment]
For others still having this problem on Debian 8, here's a somewhat workable solution:

1. Download the Stretch build of libstdc++ for your arch from one of the mirrors here: https://packages.debian.org/stretch/libstdc++6
2. Extract the .deb file (don't install). Most archive tools will let you do this as they're just standard archives.
3. Find the symlinked file libstdc++.so.6 inside the data.tar.xz archive and copy this to a suitable location on the target machine, for example /home/youruser/yourproject/libstdc++.so.6. You want the file to be named libstdc++.so.6, not libstdc++.so.6.0.22, otherwise the soname will not match and you'll still get the old libstdc++ loading.
4. Run node with the new libstdc++:
`LD_LIBRARY_PATH=/home/youruser/yourproject:$LD_LIBRARY_PATH node somescript.js`
If you have spaces in the path, you will need to escape them. Also note the lack of trailing slash in the directory name.
5. This node process will now be running with a suitably new version of libstdc++ and node-mapnik will work. All other parts of the system will still be using the standard Debian 8 libstdc++. If you are using a process manager, you can set the `LD_LIBRARY_PATH` variable in the environment vars.

For more info on this method and possible variations, see http://tldp.org/HOWTO/Program-Library-HOWTO/shared-libraries.html
For more info on potential issues with LD_LIBRARY_PATH, see http://xahlee.info/UnixResource_dir/_/ldpath.html

Someone can probably chime in on whether this is a bad idea, but for the time being it runs fine and avoids updating the whole system libstdc++ or switching distro for the sake of one node module.

**Comment by springmeyer** [comment]
This will be resolved by the upcoming node-mapnik v3.6.3 release: https://github.com/mapnik/node-mapnik/pull/839
